{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import all packages needed\n",
    "\n",
    "import os, sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import json\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Separate target data by ethnicity (ABCD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Basic setting\n",
    "\n",
    "path = '/mnt/db2/genome/ABCD'\n",
    "target = 'ABCD_release_3.0_QCed' # target file name\n",
    "demographic = '/mnt/db2/genome/ABCD/info/acspsw03.txt'\n",
    "pick = 'Others' # Choose White/Black/Hispanic/Asian/Others\n",
    "savepath = '/mnt/db2/genome/ABCD/Input/Others'\n",
    "\n",
    "PLINK = '/mnt/db2/Plink' # PLINK path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def SepEth(target, demographic, pick, savepath):\n",
    "    \n",
    "    # Set the File\n",
    "    target = pd.read_csv(path + '/' + target + '.fam', delimiter = '\\t', header = None)\n",
    "    demo = pd.read_csv(demographic, delimiter = '\\t')\n",
    "    \n",
    "    # Change type of the 'race_ethnicity' column in the demographic\n",
    "    demo = demo[['subjectkey', 'race_ethnicity']]\n",
    "    \n",
    "    # Merge two data\n",
    "    ethnicity = pd.merge(target, demo, left_on = 1, right_on = 'subjectkey')\n",
    "    ethnicity = ethnicity.drop_duplicates(['subjectkey'], keep = 'first')\n",
    "    ethnicity = ethnicity.dropna(axis=0)\n",
    "    ethnicity['race_ethnicity'] = ethnicity['race_ethnicity'].astype(int)\n",
    "    \n",
    "    # Separate ethnicity\n",
    "    if (pick == 'White'):\n",
    "        data = ethnicity[(ethnicity['race_ethnicity'] == 1)]\n",
    "    elif (pick == 'Black'):\n",
    "        data = ethnicity[(ethnicity['race_ethnicity'] == 2)]\n",
    "    elif (pick == 'Hispanic'):\n",
    "        data = ethnicity[(ethnicity['race_ethnicity'] == 3)]\n",
    "    elif (pick == 'Asian'):\n",
    "        data = ethnicity[(ethnicity['race_ethnicity'] == 4)]\n",
    "    elif (pick == 'Others'):\n",
    "        data = ethnicity[(ethnicity['race_ethnicity'] == 5)]\n",
    "\n",
    "    # Save file\n",
    "    data = data.iloc[:, :2]\n",
    "    filename = 'ABCD_' + pick\n",
    "    data.to_csv(savepath + '/' + filename + '.txt', sep = ' ', header = None, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make ethnicity file\n",
    "\n",
    "SepEth(target, demographic, pick, savepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make new binary file for ABCD separate file!\n",
    "\n",
    "os.chdir(PLINK)\n",
    "\n",
    "os.system('./plink --bfile {} --keep {} --make-bed --out {}'.format(path + '/' + target, savepath + '/' + 'ABCD_' + pick + '.txt', savepath + '/' + 'ABCD_' + pick))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quality control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "config = {\n",
    "    'Filepath' : None,\n",
    "    'Filename' : None,\n",
    "    'Bad_batch_file' : None,\n",
    "    'Phenotype_file' : None,\n",
    "    'Genotype_miss' : 0.05,\n",
    "    'Individual_miss' : 0.05,\n",
    "    'Minor allele frequency' : 0.01,\n",
    "    'Chromosome' : [1,22],\n",
    "    'Hardy weinberg equil' : 1e-6,\n",
    "    'Prune' : {\n",
    "        'window_size': 200,\n",
    "        'sliding': 50,\n",
    "        'r2': 0.25,\n",
    "    },\n",
    "    'Heterozygosity(SD)' : 3,\n",
    "    'Relatedness' : {\n",
    "        'min' : None,\n",
    "        'max' : None,\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Filepath = '/mnt/db2/genome/ABCD/Input/Others'\n",
    "Filename = 'ABCD_Others'\n",
    "\n",
    "PLINK = '/mnt/db2/Plink' # PLINK path\n",
    "\n",
    "Bad_batch_file = 'batch_bad.txt'\n",
    "\n",
    "Phenotype_file = None #'qtrait.txt'\n",
    "\n",
    "# Set the thresholds of --geno, --mind, --maf --hwe\n",
    "Genotype_miss = None\n",
    "Individual_miss = None\n",
    "MAF = None\n",
    "HWE = None\n",
    "\n",
    "Chromosome = None #[,]\n",
    "\n",
    "Prune_window_size = None\n",
    "Prune_sliding = None\n",
    "Prune_r2 = None\n",
    "\n",
    "Heterozygosity_SD = 4\n",
    "\n",
    "Relatedness_min = 0.4\n",
    "Relatedness_max = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Config\n",
    "config['Filepath'] = Filepath\n",
    "config['Filename'] = Filename\n",
    "config['Bad_batch_file'] = Bad_batch_file\n",
    "config['Genotype_miss'] = config['Genotype_miss'] if Genotype_miss == None else Genotype_miss\n",
    "config['Individual_miss'] = config['Individual_miss'] if Individual_miss == None else Individual_miss\n",
    "config['Minor allele frequency'] = config['Minor allele frequency'] if MAF == None else MAF\n",
    "config['Chromosome'] = config['Chromosome'] if Chromosome == None else Chromosome\n",
    "config['Prune']['window_size'] = config['Prune']['window_size'] if Prune_window_size == None else Prune_window_size\n",
    "config['Prune']['sliding'] = config['Prune']['sliding'] if Prune_sliding == None else Prune_sliding\n",
    "config['Prune']['r2'] = config['Prune']['r2'] if Prune_r2 == None else Prune_r2\n",
    "config['Heterozygosity(SD)'] = config['Heterozygosity(SD)'] if Heterozygosity_SD == None else Heterozygosity_SD\n",
    "config['Relatedness']['min'] = Relatedness_min\n",
    "config['Relatedness']['max'] = Relatedness_max"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make all directory\n",
    "\n",
    "os.mkdir(config['Filepath'] + '/' + \"Final_results\")\n",
    "os.mkdir(config['Filepath'] + '/' + '1.Bad_batch_removed')\n",
    "os.mkdir(config['Filepath'] + '/' + '2.Missing_genotype')\n",
    "os.mkdir(config['Filepath'] + '/' + '3.Missing_individuals')\n",
    "os.mkdir(config['Filepath'] + '/' + '4.Impute_sex')\n",
    "os.mkdir(config['Filepath'] + '/' + '5.Choose_chromosome')\n",
    "os.mkdir(config['Filepath'] + '/' + '6.Minor_allele_frequency')\n",
    "os.mkdir(config['Filepath'] + '/' + '7.HWE')\n",
    "os.mkdir(config['Filepath'] + '/' + '8.Pruning')\n",
    "os.mkdir(config['Filepath'] + '/' + '9.Heterozygosity')\n",
    "os.mkdir(config['Filepath'] + '/' + '10.Relatedness')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change direcctory which PLINK software is in\n",
    "os.chdir(PLINK)\n",
    "\n",
    "#Save config\n",
    "with open(config['Filepath'] + '/' + 'Final_results/Used_parameters.json', 'w') as f:\n",
    "        json.dump(config, f,indent = 0)\n",
    "\n",
    "#Exclude bad batch\n",
    "if config['Bad_batch_file'] != None:\n",
    "    print('Bad batch file detected! Processing...')\n",
    "    os.system('./plink --bfile {} --remove {} --out {}/1.Bad_batch_removed/ABCD_excluded --make-bed'.format(config['Filepath'] + '/' +  config['Filename'], config['Filepath'] + '/' + config['Bad_batch_file'], config['Filepath']))\n",
    "else:\n",
    "    print('No bad batch file detected. Anyway, New file will be made!')\n",
    "    os.system('./plink --bfile {} --out {}/1.Bad_batch_removed/ABCD_excluded --make-bed'.format(config['Filepath'] + '/' + config['Filename'], config['Bad_batch_file'], config['Filepath']))\n",
    "\n",
    "#Exclude missing genotype\n",
    "os.system('./plink --bfile {}/1.Bad_batch_removed/ABCD_excluded --geno {} --out {}/2.Missing_genotype/ABCD_geno --make-bed'.format(config['Filepath'], config['Genotype_miss'], config['Filepath']))\n",
    "\n",
    "#Exclude missing individuals\n",
    "os.system('./plink --bfile {}/2.Missing_genotype/ABCD_geno --mind {} --out {}/3.Missing_individuals/ABCD_mind --make-bed'.format(config['Filepath'], config['Individual_miss'], config['Filepath']))\n",
    "\n",
    "#Impute sex\n",
    "os.system('./plink --bfile {}/3.Missing_individuals/ABCD_mind --impute-sex --out {}/4.Impute_sex/ABCD_sex --make-bed'.format(config['Filepath'], config['Filepath']))\n",
    "\n",
    "#Choose which chromosome to use\n",
    "os.system('./plink --bfile {}/4.Impute_sex/ABCD_sex --chr {}-{} --out {}/5.Choose_chromosome/ABCD_chr --make-bed'.format(config['Filepath'], config['Chromosome'][0],\n",
    "                                                                                                                  config['Chromosome'][1], config['Filepath']))\n",
    "\n",
    "#Minor Allele Frequency\n",
    "os.system('./plink --bfile {}/5.Choose_chromosome/ABCD_chr --maf {} --out {}/6.Minor_allele_frequency/ABCD_maf --make-bed'.format(config['Filepath'], config['Minor allele frequency'], config['Filepath']))\n",
    "\n",
    "#Hardy Weinberg Equilibrium\n",
    "os.system('./plink --bfile {}/6.Minor_allele_frequency/ABCD_maf --hwe {} --out {}/7.HWE/ABCD_hwe --make-bed'.format(config['Filepath'], config['Hardy weinberg equil'], config['Filepath']))\n",
    "\n",
    "#Pruning\n",
    "os.system('./plink --bfile {}/7.HWE/ABCD_hwe --indep-pairwise {} {} {} --out {}/8.Pruning/ABCD_pruned'.format(config['Filepath'], config['Prune']['window_size'],\n",
    "                                                                                       config['Prune']['sliding'],\n",
    "                                                                                       config['Prune']['r2'], config['Filepath']))\n",
    "\n",
    "#Make pruned dataset\n",
    "os.system('./plink --bfile {}/7.HWE/ABCD_hwe --extract {}/8.Pruning/ABCD_pruned.prune.in --out {}/8.Pruning/ABCD_pruned --make-bed'.format(config['Filepath'], config['Filepath'], config['Filepath']))\n",
    "\n",
    "#Heterozygosity with pruned dataset\n",
    "os.system('./plink --bfile {}/8.Pruning/ABCD_pruned --het --out {}/9.Heterozygosity/ABCD_het'.format(config['Filepath'], config['Filepath']))\n",
    "\n",
    "#Make removal list with heterozygosity calculation\n",
    "ABCD_het = pd.read_csv(config['Filepath'] + '/' + '9.Heterozygosity/ABCD_het.het', sep = ' ', low_memory = False,\n",
    "                       skipinitialspace=True)\n",
    "\n",
    "#Calculating heterozygosity\n",
    "ABCD_het['het'] = (ABCD_het['N(NM)']-ABCD_het['O(HOM)'])/ABCD_het['N(NM)']\n",
    "mean = np.mean(ABCD_het['het'])\n",
    "std = np.std(ABCD_het['het'])\n",
    "\n",
    "over_out = ABCD_het.loc[ABCD_het['het'] > mean+int('{}'.format(config['Heterozygosity(SD)']))*std,:]\n",
    "under_out = ABCD_het.loc[ABCD_het['het'] < mean-int('{}'.format(config['Heterozygosity(SD)']))*std, :]\n",
    "remove = pd.concat([over_out,under_out], axis=0)\n",
    "remove.to_csv(config['Filepath'] + '/' + '9.Heterozygosity' +'/' + 'het_remove.txt', sep = ' ', header=None, index = False)\n",
    "\n",
    "os.system('./plink --bfile {}/7.HWE/ABCD_hwe --remove {}/9.Heterozygosity/het_remove.txt --out {}/9.Heterozygosity/ABCD_het_removed --make-bed'.format(config['Filepath'], config['Filepath'], config['Filepath']))\n",
    "\n",
    "if config['Relatedness']['min'] != None and config['Relatedness']['max'] == None:\n",
    "    os.system('./plink --bfile {}/8.Pruning/ABCD_pruned --genome --min {} --out {}/10.Relatedness/ABCD_genome'.format(config['Filepath'], config['Relatedness']['min'], config['Filepath']))\n",
    "if config['Relatedness']['min'] == None and config['Relatedness']['max'] != None:\n",
    "    os.system('./plink --bfile {}/8.Pruning/ABCD_pruned --genome --max {} --out {}/10.Relatedness/ABCD_genome'.format(config['Filepath'], config['Relatedness']['max'], config['Filepath']))\n",
    "if config['Relatedness']['min'] != None and config['Relatedness']['max'] != None:\n",
    "    os.system('./plink --bfile {}/8.Pruning/ABCD_pruned --genome --min {} --max {} --out {}/10.Relatedness/ABCD_genome'.format(config['Filepath'], config['Relatedness']['min'],\n",
    "                                                                                                         config['Relatedness']['max'], config['Filepath']))\n",
    "if config['Relatedness']['min'] == None and config['Relatedness']['max'] == None:\n",
    "    os.system('./plink --bfile {}/8.Pruning/ABCD_pruned --genome --out {}/10.Relatedness/ABCD_genome'.format(config['Filepath'], config['Filepath']))\n",
    "\n",
    "ABCD_genome = pd.read_csv(config['Filepath'] + '/' + '10.Relatedness/ABCD_genome.genome', sep = ' ', low_memory = False,\n",
    "                       skipinitialspace=True)\n",
    "\n",
    "#get missing rate information\n",
    "os.system('./plink --bfile {}/1.Bad_batch_removed/ABCD_excluded --missing --out {}/10.Relatedness/ABCD_miss'.format(config['Filepath'], config['Filepath']))\n",
    "miss = pd.read_csv(config['Filepath'] + '/' + '10.Relatedness/ABCD_miss.imiss', sep = ' ', low_memory = False,\n",
    "                       skipinitialspace=True)\n",
    "\n",
    "#Compare individuals missing rate and leave only good one for each pairs\n",
    "low_call_list = []\n",
    "low_call_val = []\n",
    "for i in range(ABCD_genome.shape[0]):\n",
    "    f1 = ABCD_genome.loc[i,'FID1']\n",
    "    p1 = ABCD_genome.loc[i,'IID1']\n",
    "    f2 = ABCD_genome.loc[i,'FID2']\n",
    "    p2 = ABCD_genome.loc[i,'IID2']\n",
    "    c1 = float(miss.loc[miss['IID'] == p1, 'F_MISS'])\n",
    "    c2 = float(miss.loc[miss['IID'] == p2, 'F_MISS'])\n",
    "    if c1>c2:\n",
    "        low_call_list.append([f1,p1])\n",
    "        low_call_val.append([c1,c2])\n",
    "    elif c2>c1:\n",
    "        low_call_list.append([f2,p2])\n",
    "        low_call_val.append([c1,c2])\n",
    "    else: continue\n",
    "        \n",
    "#Save low call individual list\n",
    "f = open(config['Filepath'] + '/' + '10.Relatedness' + '/' + \"low_call.txt\", 'w')\n",
    "for i in range(len(low_call_list)):\n",
    "    a1 = str(low_call_list[i]).replace(\"['\",'')\n",
    "    a2 = a1.replace(\"']\",'')\n",
    "    a3 = a2.replace(\"', '\",' ')\n",
    "    data = a3 + '\\n'\n",
    "    f.write(data)\n",
    "f.close()\n",
    "\n",
    "#Exclude low call list from data\n",
    "os.system('./plink --bfile {}/9.Heterozygosity/ABCD_het_removed --remove {}/10.Relatedness/low_call.txt --out {}/Final_results/ABCD_QC_finished --make-bed'.format(config['Filepath'], config['Filepath'], config['Filepath']))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
